enterprise/backend/src/metabase_enterprise/workspaces/models/workspace_log.clj

```diff
@@ -0,0 +1,54 @@
+(ns metabase-enterprise.workspaces.models.workspace-log
+  (:require
+   [java-time.api :as t]
+   [metabase.models.interface :as mi]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]))
+
+(methodical/defmethod t2/table-name :model/WorkspaceLog [_model] :workspace_log)
+
+(doto :model/WorkspaceLog
+  (derive :metabase/model))
+
+(t2/deftransforms :model/WorkspaceLog
+  {:task   mi/transform-keyword
+   :result mi/transform-keyword})
+
+(defn start!
+  "Create a log entry for a workspace setup task. Returns the created log."
+  [workspace-id task]
+  (t2/insert-returning-instance! :model/WorkspaceLog
+                                 {:workspace_id workspace-id
+                                  :task         task
+                                  :started_at   (t/offset-date-time)
+                                  :result       :started}))
+
+(defn- complete!
+  [result log-id & [message]]
+  (assert (#{:success :failure} result) "It can only be either success or failure")
+  (let [data (cond-> {:completed_at (t/offset-date-time)
+                      :result       result}
+               message (assoc :message message))]
+    (t2/update! :model/WorkspaceLog log-id data)))
+
+(defn success!
+  "Task was completed successfully"
+  [log-id & [message]]
+  (complete! :success log-id message))
+
+(defn failure!
+  "Task has failed"
+  [log-id & [message]]
+  (complete! :failure log-id message))
+
+(defmacro track!
+  "Wrap body with automatic logging: starts a log entry, marks success on completion, failure on exception."
+  [workspace-id task & body]
+  `(let [log# (start! ~workspace-id ~task)]
+     (try
+       (let [result# (do ~@body)]
+         (success! (:id log#))
+         result#)
+       (catch Exception e#
+         (failure! (:id log#) (ex-message e#))
+         (throw e#)))))
```

Comment:
Watch out that these updates are actually flushed so the FE can read them.

Could dynamically check that this stuff isn't run in a transaction, or have it manually commit. 

Both of these are pretty gross 🤷